### ISSUE TITLE
Improved cleaning of domains.list and domains-dotted-format.list

### ISSUE THREAD

**Issue by mitchellkrogza** [mitchellkrogza/Ultimate.Hosts.Blacklist#93]
Thanks so much to @maravento for all his continuous help with this project. 

Over a period of weeks he has helped me made numerous improvements to cleaning the lists of invalid characters and today we have finally reached that point. Most of this was discussed in https://github.com/mitchellkrogza/Ultimate.Hosts.Blacklist/issues/72 but I decided to start this as new topic instead.

This morning after a few hours of tweaking all the sed and awk commands that @marvento has contributed to the project I have finally perfected the creation of 2 very clean lists for use in Pi-Hole's and Dnsmasq systems.

The two files domains.list and domains-dotted-format.list and now void of the following:

- Any domain that has a string attached ie. `domain.com/something/blog/index.php` is now just `domain.com` or `.domain.com` depending on which list we refer to.

- Any beginning of a domain name that contains any underscore `_` characters. The underscore character is illegal and invalid in any DNS system anywhere in the world. So domains like `10000_gratis_sexfoto.jouwpagina.nl` will now just be `jouwpagina.nl` or `.jouwpagina.nl` - similarly a domain like `___id___.c.mystat-in.net` now becomes `c.mystat-in.net` or `.c.mystat-in.net`

- Any domains beginning with `www.`, `ftp.`, `ww.`, `zzz.` are now also stripped out. HOWEVER for the hosts file they will not be stripped out due to a hosts file not being able to do wildcarding.

No doubt we will be tweaking and perfecting this even more over time but we are getting ever closer to producing the cleanest hosts file anywhere out there and we are also able to show how many input sources out there are VERY BADLY maintained as they simply add anything to their hosts files without ever checking them, nor do they ever clean their lists. Anyone who knows anything should at least know that underscore characters are illegal and invalid in DNS but clearly the people producing hosts blocker files have no idea about simple networking principles and fill their lists with utter junk.

Forwards and Upwards we move !!!

**Comment by mitchellkrogza** [comment]
A few extra commits and tweaks and the lists are looking almost perfect. @maravento, have a look later and let me know what you think.

**Comment by maravento** [comment]
give me the link

**Comment by mitchellkrogza** [comment]
https://hosts.ubuntu101.co.za/domains.list

and

https://hosts.ubuntu101.co.za/domains-dotted-format.list

**Comment by maravento** [comment]
domains-dotted-format.list:
.www.klimexs.com
.www8.paypopup.com
.www89.chunguang100.com
.www8-ssl.effectivemeasure.net
.www99.zapto.org
.wwwa024.infonegocio.com
.www99*+.mytest.com  # This is a domain invented by me, only for testing purposes

sed -r  's:(^\.*?(www|ftp)[^.]*?\.|^\.\.?)::gi' domains-dotted-format.list | awk '{print "."$1}' | sort -u > newlist

newlist:
.chunguang100.com
.effectivemeasure.net
.infonegocio.com
.klimexs.com
.mytest.com
.paypopup.com
.zapto.org

**Comment by mitchellkrogza** [comment]
Thanks @maravento I will double check in the morning and we will get this perfect. Had a problem with one of your sed’s which was taking the domains at the top of the file which are like 0———-0———-.com and stripping the 0 off. Could of been something I did wrong but I’ll test again tomorrow and get it perfect. Thanks again.

**Comment by mitchellkrogza** [comment]
@maravento please download those same two files again now and see if you see a difference. Discovered a problem with the way I update the raw files before the commit has actually occurred. Let me know.

**Comment by maravento** [comment]
The solution is only to admit domains that contain letters, numbers or hyphens

list
.www99*+-_.mytest.com
.WWW01*+-_22.mytest2.com
.0—�—�—�-0—�—�—�-.com
.test.com

sed -r '/[^a-zA-Z0-9.-]/d' list

test.com

**Comment by mitchellkrogza** [comment]
Sorry @maravento I typed that message on my iPhone and it really messed things up. Can you believe the iPhone keyboard doesn't have a backtick, either that or I am blind.

Let me try again.

Domains like
```
0----0.0----0.1596.hk
0--0.0--0.hnun.net
0------------0-------------0.0n-line.info
```

Were being stripped like this 

```
----0.0----0.1596.hk
--0.0--0.hnun.net
------------0-------------0.0n-line.info
```

which has been resolved in my latest lists but I still need to address removing the www. www99. ftp. etc which I will test with your latest recommendation.

**Comment by maravento** [comment]
sed '/0--/d'

**Comment by mitchellkrogza** [comment]
Thanks @maravento I will try that, let me expand a bit as this is getting complex for me.

Input list

```
0------------0-------------0.0n-line.info
0-0--0-000.com
0-3.us
aw.dermalmask.com
idolstudio.free.fr
idolstudio2.free.fr
something.blogspot.com
anything.blogspot.com
xxx.blogspot.ca
www.hola.org
www10.a8.net
www11.alsto.com
www148.myquicksearch.com
ftp.thaitattoo.nl
ftp01.pornocrawler.ws
ftp04.pornocrawler.ws
g.blogads.com
wvw.tielecreidito-pe.com
```

Output should be

```
0n-line.info
0-0--0-000.com
0-3.us
dermalmask.com
idolstudio.free.fr
idolstudio2.free.fr
something.blogspot.com
anything.blogspot.com
xxx.blogspot.ca
hola.org
a8.net
alsto.com
myquicksearch.com
thaitattoo.nl
pornocrawler.ws
pornocrawler.ws
blogads.com
tielecreidito-pe.com
```

**Comment by mitchellkrogza** [comment]
@maravento I tried this again and this seems very close to what we need `sed -r '/[^a-zA-Z0-9.-]/d' input.txt | sed -r 's:(^.?(www|ftp)[^.]?.|^..?)::gi' | awk '{print "."$1}' | sed -r '/^.\W+/d' | sort -u > output.txt`

But still a domain like

`0------------0-------------0.0n-line.info`

should be left as 

`0n-line.info`

But it's stripped completely from the list.

**Comment by maravento** [comment]
sed '/0--/d' input | sed -r  's:(^\.*?(www|ftp|xxx|wvw)[^.]*?\.|^\.\.?)::gi' | awk '{print "."$1}' | sed -r '/^\.\W+/d' | sort -u > output

.0-3.us
.a8.net
.alsto.com
.anything.blogspot.com
.aw.dermalmask.com
.blogspot.ca
.g.blogads.com
.hola.org
.idolstudio2.free.fr
.idolstudio.free.fr
.myquicksearch.com
.pornocrawler.ws
.something.blogspot.com
.thaitattoo.nl
.tielecreidito-pe.com

**Comment by mitchellkrogza** [comment]
Thanks @maravento tried that but still `0------------0-------------0.0n-line.info` and domain like `idontgiveafuck.com` completely stripped out. 🙄 this is getting ever more complex

**Comment by maravento** [comment]
input:
0------------0-------------0.0n-line.info
test.com
sed '/0--/d' input > output
output:
test.com

**Comment by mitchellkrogza** [comment]
@maravento I think you have nailed it I did this in this order

```
sed '/0--/d' domains-in.txt  > domains-out.txt
sed -r 's:(^.?(www|ftp|xxx|wvw)[^.]?.|^..?)::gi' domains-out.txt | awk '{print "."$1}' | sed -r '/^.\W+/d' | sort -u > domains-dotted.txt
```

**Comment by mitchellkrogza** [comment]
Soooo close, we are still ending up with these in the list (just a few examples there are many many)

```
.www13.glam.com
.www146.lewwwz.com
.www1.scat-sex.org
.www.zootravel.com
.www.ftp.thaitattoo.nl
.ftp.track4.com
```

**Comment by maravento** [comment]
I do not understand what happens with your list. Example:
input:
.www13.glam.com
.www146.lewwwz.com
.www1.scat-sex.org
.www.zootravel.com
.www.ftp.thaitattoo.nl
.ftp.track4.com

sed -r 's:(^.?(www|ftp|xxx|wvw)[^.]?.|^..?)::gi' input

glam.com
lewwwz.com
scat-sex.org
zootravel.com
ftp.thaitattoo.nl
track4.com

you can manually delete this line ("ftp.thaitattoo.nl") or run the command twice

**Comment by maravento** [comment]
complete command:
sed '/[A-Z]/d' **oldlist** | sed '/0--/d' | sed -r '/[^a-zA-Z0-9.-]/d' | sed -r 's:(^.?(www|ftp|xxx|wvw)[^.]?.|^..?)::gi'  | awk '{print "."$1}' | sed -r '/^\.\W+/d' | sort -u > **newlist**

**Comment by mitchellkrogza** [comment]
Thanks @maravento I'll try that in the morning .... my head is all sed'd and awk'd out for one day 😁 🤣

**Comment by funilrys** [comment]
Closing for now feel free to reopen once you reviewed the new system @mitchellkrogza :+1:
